chalk-solve/src/clauses/builtin_traits/pointee.rs

```diff
@@ -0,0 +1,145 @@
+    let mut push = |m| push_clauses(db, builder, self_ty.clone(), m);
+    match self_ty.kind(interner) {
+        TyKind::Str | TyKind::Slice(_) => {
+            push(TyKind::Scalar(chalk_ir::Scalar::Uint(chalk_ir::UintTy::Usize)).intern(interner))
+        }
+        TyKind::Array(_, _)
+        | TyKind::Never
+        | TyKind::Closure(_, _)
+        | TyKind::FnDef(_, _)
+        | TyKind::Scalar(_)
+        | TyKind::Raw(_, _)
+        | TyKind::Function(_)
+        | TyKind::InferenceVar(_, TyVariableKind::Float)
+        | TyKind::InferenceVar(_, TyVariableKind::Integer)
+        | TyKind::Generator(_, _)
+        | TyKind::GeneratorWitness(_, _)
+        | TyKind::Ref(_, _, _) => {
+            push(TyKind::Tuple(0, Substitution::empty(interner)).intern(interner))
+        }
+        TyKind::Adt(id, subst) => {
+            let adt_datum = db.adt_datum(*id);
+            if adt_datum.kind != AdtKind::Struct {
+                push(TyKind::Tuple(0, Substitution::empty(interner)).intern(interner));
+                return Ok(());
+            }
+            let last_field_ty = adt_datum
+                .binders
+                .map_ref(|b| b.variants.clone())
+                .substitute(interner, subst)
+                .into_iter()
+                .take(1) // We have a struct so we're guaranteed one variant
+                .flat_map(|mut v| v.fields.pop())
+                .next();
+            push_for_last_field(last_field_ty, db, builder, self_ty);
+        }
+        TyKind::Tuple(_, subst) => {
+            let last_field_ty = subst
+                .iter(interner)
+                .rev()
+                .next()
+                .and_then(|x| x.ty(interner))
+                .cloned();
+            push_for_last_field(last_field_ty, db, builder, self_ty);
+        }
+        TyKind::BoundVar(_)
+        | TyKind::AssociatedType(_, _)
+        | TyKind::OpaqueType(_, _)
+        | TyKind::Foreign(_)
+        | TyKind::Error
+        | TyKind::Placeholder(_)
+        | TyKind::Dyn(_)
+        | TyKind::Alias(_) => (),
+        TyKind::InferenceVar(_, TyVariableKind::General) => return Err(Floundered),
+    }
+    Ok(())
+}
+
+fn push_for_last_field<I: Interner>(
+    last_field_ty: Option<Ty<I>>,
+    db: &dyn RustIrDatabase<I>,
+    builder: &mut ClauseBuilder<'_, I>,
+    self_ty: Ty<I>,
+) {
+    let interner = db.interner();
+    if let Some(last_field_ty) = last_field_ty {
+        let _ = add_pointee_program_clauses(db, builder, last_field_ty.clone());
+        let trait_id = db.well_known_trait_id(WellKnownTrait::Pointee).unwrap();
+        let trait_datum = db.trait_datum(trait_id);
+        assert_eq!(
+            trait_datum.associated_ty_ids.len(),
+            1,
+            "Pointee trait should have exactly one associated type, found {:?}",
+            trait_datum.associated_ty_ids
+        );
+        let metadata_id = trait_datum.associated_ty_ids[0];
+        builder.push_fact(TraitRef {
```

Comment:
Perhaps we should just push this trait clause at the top of `add_pointee_program_clauses` instead of in `push_clauses` and here. We still want trait predicates like `dyn Trait: Pointee` to hold, even if we can't normalize the associated type to anything useful currently?